eng/tools/spec-gen-sdk-runner/src/utils.ts

```diff
@@ -94,5 +132,209 @@ export function runPowerShellScript(args: string[]): string | undefined {
+      result[lastFolder].push(filePath);
+    }
+    const relatedFolder = searchRelatedFolder(filePath, options);
+    if (relatedFolder === undefined) {
+      continue;
+    }
+    if (result[relatedFolder] === undefined) {
+      result[relatedFolder] = [];
+    }
+    result[relatedFolder].push(filePath);
+    lastFolder = relatedFolder;
+  }
+
+  return result;
+};
+
+/*
+ * Search for the related type spec projects from a shared library
+ */
+export const searchRelatedTypeSpecProjectBySharedLibrary = async (
+  sharedLibraries: { [relatedFolder: string]: string[] },
+  options: FsSearchOptions,
+) => {
+  const result: { [relatedFolder: string]: string[] } = {};
+  for (const sharedLibrary of Object.keys(sharedLibraries)) {
+    const parentFolder = path.dirname(sharedLibrary);
+    const fileNames = await getFilesInFolder(parentFolder, options);
+    for (const fileName of fileNames) {
+      const filePath = path.join(parentFolder, fileName);
+      const subFileNames = await getFilesInFolder(filePath, options);
+      for (const subFileName of subFileNames) {
+        if (options.searchFileRegex.test(subFileName)) {
+          if (!result[filePath]) {
+            result[filePath] = [];
+          }
+          result[filePath] = [...result[filePath], ...sharedLibraries[sharedLibrary]];
+        }
+      }
+    }
+  }
+  return result;
+};
+
+/**
+ * Search from the parent folders for a list of files
+ */
+export const searchRelatedParentFolders = async (fileList: string[], options: FsSearchOptions) => {
+  const result: { [relatedFolder: string]: string[] } = {};
+  fileList.sort();
+
+  for (const filePath of fileList) {
+    const relatedParentFolder = await searchRelatedParentFolder(filePath, options);
+    if (relatedParentFolder === undefined) {
+      continue;
+    }
+    if (result[relatedParentFolder] === undefined) {
+      result[relatedParentFolder] = [];
+    }
+    result[relatedParentFolder].push(filePath);
+  }
+
+  return result;
+};
+
+/*
+ * Search from a nearest parent folder for the specific file pattern
+ */
+export const searchRelatedParentFolder = async (filePath: string, options: FsSearchOptions) => {
+  let searchPath = filePath;
+
+  while (searchPath !== ".") {
+    const fileNames = await getFilesInFolder(searchPath, options);
+    for (const fileName of fileNames) {
+      if (options.searchFileRegex.test(fileName)) {
+        return searchPath;
+      }
+    }
+    searchPath = path.dirname(searchPath);
+  }
+
+  return undefined;
+};
+
+/*
+ * Get the files in a folder
+ */
+const getFilesInFolder = async (
+  searchPath: string,
+  options: FsSearchOptions,
+): Promise<string[]> => {
+  const workingFolder = ".";
+  const workPath = path.resolve(process.cwd(), workingFolder, options.specFolder, searchPath);
+  // Execute the git command using exec
+  const { stdout, stderr } = await execAsync(`git ls-tree ${options.treeId} ${workPath}`);
```

Comment:
Why are we use a git command here instead of just traversing the file system? I would expect by the name of this function we are only looking at the files on disk.

Reply:
Updated the code to use file system access directly.